### PR TITLE
git: Fix submodule fetch, rewrite logic

### PR DESCRIPTION
**Pull request by ReillyBrogan**
The logic for this was quite broken and had several failure cases due to that. This commit re-writes that to be saner.

Changes:
- Individual exec commands are all split out to separate functions so that they can be more easily composed. No function does more than a single command
- Control flow now happens entirely in the Fetch() function
- Before there were several failure states around submodules due to the `git switch` command happening immediately after the `git fetch` if the repo was already cloned. This is fixed by moving to a unified control flow.
- `Fetch()` will now run either the initial clone OR update refs before moving onto further steps
- `git clone` no longer attempts to initialize submodules. Several features from `git submodules` aren't yet implemented by `git clone` and so submodule initialization is moved entirely into a `git submodule` command
- `git switch` now happens after the clone/fetch but before `git submodule`
- `git switch` now no longer runs with any submodule-related flags. Like `git clone` the features we need are just not implemented yet and as far as I can tell there are subtle behavioral differences that were causing issues.
- `git submodule` will now use `--filter=blob:none` to specify that submodule checkouts should themselves be blobless when possible. As far as I can tell this works correctly with `--recursive` and recursive submodules.
- Because we are now checking out submodules with blobless clones when possible the `fixPermissions()` logic has been updated to account for promisor files that could be in submodule `.git` directories

### CODE REVIEW

**Comment by ReillyBrogan** [comment]
> Comment should end in a period

I don't tell _you_ how to live your life CI

File: builder/source/git.go [hunk 1]
```diff
@@ -65,90 +65,85 @@ func NewGit(uri, ref string) (*GitSource, error) {
...
-	fetchCmd := exec.Command("git", fetchArgs...)
-	fetchCmd.Dir = path
-
-	if err := fetchCmd.Run(); err != nil {
-		return err
-	}
+// For some reason git blobless clones create pack files with 600 permissions. These break future operations
+// as those files cannot be read by non-root users. Fix those permissions so things work as they should.
+func (g *GitSource) fixPermissions() error {
+	cmd := exec.Command("bash", "-c", "find . -name \"*.promisor\" -type f -exec chmod +r \"{}\" \\;")
```
Comment (GZGavinZhao) [inline builder/source/git.go]: Slight nit, might want to use raw string for the `find` command?

**Review by GZGavinZhao — DISMISSED** [review]
Logic is sane to me. If this has been tested, then feel free to go ahead and merge!

**Review by GZGavinZhao — APPROVED** [review]

**Review by silkeh — APPROVED** [review]
LGTM, thanks!

**Pull request merged**